File: LinkedList/143. Reorder List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

//Solution 1 (NeetCode)

class Solution {
public:
    void reorderList(ListNode* head) {
        ListNode* slow=head;
        ListNode* fast=head->next;
        while (fast!=NULL && fast->next!=NULL) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        ListNode* second=slow->next;
        ListNode* first=head;
        slow->next = NULL;
        
        second = reverse(second);
        merge(first, second);
    }

private:
    ListNode* reverse(ListNode* node) {
        ListNode* prev=NULL;
        ListNode* tmp;
        while (node != NULL) {
            // ListNode* tmp = node->next;
            tmp = node->next;
            node->next = prev;
            prev = node;
            node = tmp;
        }
        return prev;
    }
    
    
    void merge(ListNode* l1, ListNode* l2) {
        while (l2 != NULL) {
            ListNode* tmp1=l1->next;
            ListNode* tmp2=l2->next;
            l1->next = l2;
            l2->next = tmp1;
            // l1, l2 = tmp1, tmp2;  //why這樣寫會超過time limit？
            l1 = tmp1;
            l2 = tmp2;
        }
    }
};

File: 06.LinkedList/143. Reorder List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

### Solution1 (NeetCode) ###

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow, fast = head, head.next
        
        while fast and fast.next:  #因為長度可能為偶數or奇數
            slow = slow.next
            fast = fast.next.next
        
        second = slow.next
        prev = slow.next = None
        while second:
            tmp = second.next
            second.next = prev
            prev = second
            second = tmp
        
        first, second = head, prev
        # while first:  #在奇數長度時會出錯
        while second:   #奇偶數長度均可以
            tmp1, tmp2 = first.next, second.next
            first.next = second
            second.next = tmp1
            first, second = tmp1, tmp2

